<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>snakemakelib_workflows.scrnaseq.app.pca &#8212; snakemakelib_workflows 0.1-alpha.9+1.g63cbc92 documentation</title>
    
    <link rel="stylesheet" href="../../../../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/bootstrap-3.3.4/css/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/bootstrap-3.3.4/css/bootstrap-theme.min.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../../',
        VERSION:     '0.1-alpha.9+1.g63cbc92',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../../../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../../../../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../../../../_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../../../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
    <link rel="top" title="snakemakelib_workflows 0.1-alpha.9+1.g63cbc92 documentation" href="../../../../index.html" />
    <link rel="up" title="Module code" href="../../../index.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body role="document">


<div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">

    <div class="container">

      <div class="navbar-header">

        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>

        <a class="navbar-brand" href="../../../../index.html">snakemakelib_workflows</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1-alpha.9+1.g63cbc92</b></span>
      </div>
        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../../../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/workflows.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/workflows.html#general-comments">General comments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/workflows.html#workflow-list">Workflow list</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../docs/workflows/atacseq.html">ATAC-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/atacseq.html#configuration">Configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/atacseq.html#workflow">Workflow</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/atacseq.html#troubleshooting">Troubleshooting</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/atacseq.html#reference-guide">Reference guide</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html">Single cell RNA-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#required-configuration">Required configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#workflow-specific-configuration">Workflow specific configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#other-configuration">Other configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#additional-advice">Additional advice</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#workflow">Workflow</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#troubleshooting">Troubleshooting</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#reference-guide">Reference guide</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/workflows.html#known-issues">Known issues</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/workflows.html#troubleshooting">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/reference.html">Reference Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../docs/reference/atacseq.html"><code class="docutils literal"><span class="pre">snakemakelib_workflows.atacseq.app</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/atacseq.html#graphics"><code class="docutils literal"><span class="pre">graphics</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/atacseq.html#hooks"><code class="docutils literal"><span class="pre">hooks</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/atacseq.html#results"><code class="docutils literal"><span class="pre">results</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html"><code class="docutils literal"><span class="pre">snakemakelib_workflows.scrnaseq.app</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#graphics"><code class="docutils literal"><span class="pre">graphics</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#hooks"><code class="docutils literal"><span class="pre">hooks</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#pca"><code class="docutils literal"><span class="pre">pca</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#report"><code class="docutils literal"><span class="pre">report</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#technicalnoise"><code class="docutils literal"><span class="pre">technicalnoise</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#utils"><code class="docutils literal"><span class="pre">utils</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/release_notes.html">Release notes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../docs/release_notes.html#pre-releases">Pre-releases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.9.html">0.1-alpha.9 (Oct 11 2016)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.8.html">0.1-alpha.8 (Jan 8 2016)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.7.html">0.1-alpha.7 (Jan 7 2016)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.6.html">0.1-alpha.6 (Dec 13 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.5.html">0.1-alpha.5 (Dec 12 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.4.html">0.1-alpha.4 (Dec 12 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.3.html">0.1-alpha.3 (Dec 12 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.2.html">0.1-alpha.2 (Nov 24 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.1.html">0.1-alpha.1 (Oct 8 2015)</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"></ul>
</li>
              
            
            
            
            
            
          </ul>

        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <h1>Source code for snakemakelib_workflows.scrnaseq.app.pca</h1><div class="highlight"><pre>
<span class="c"># Copyright (C) 2015 by Per Unneberg</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">snakemakelib.log</span> <span class="k">import</span> <span class="n">LoggerManager</span>

<span class="n">smllogger</span> <span class="o">=</span> <span class="n">LoggerManager</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="n">__name__</span><span class="p">)</span>

<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;scrnaseq_pca_all&#39;</span><span class="p">]</span>

<div class="viewcode-block" id="scrnaseq_pca_all"><a class="viewcode-back" href="../../../../docs/reference/scrnaseq.html#snakemakelib_workflows.scrnaseq.app.pca.scrnaseq_pca_all">[docs]</a><span class="k">def</span> <span class="nf">scrnaseq_pca_all</span><span class="p">(</span><span class="n">expr</span><span class="p">,</span> <span class="n">pca_output_file</span><span class="p">,</span> <span class="n">pcaobj_pickle_file</span><span class="p">,</span>
                     <span class="n">metadata</span><span class="o">=</span><span class="k">None</span><span class="p">,</span> <span class="n">pca_quant</span><span class="o">=</span><span class="s">&quot;TPM&quot;</span><span class="p">,</span>
                     <span class="n">screen_n</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Wrapper for running pca and feature selection.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">      expr (str): input file name</span>
<span class="sd">      pca_output_file (str): output file name</span>
<span class="sd">      pcaobj_pickle_file (str): output file name for pickle serialization</span>
<span class="sd">      metadata (str): metadata file name</span>
<span class="sd">      screen_n (int): number of features to keep based on variance</span>
<span class="sd">      </span>
<span class="sd">    Returns:</span>
<span class="sd">      None - results are saved to given output files</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s">&quot;samples&quot;</span><span class="p">,</span> <span class="k">None</span><span class="p">)</span>
    <span class="n">expr_long</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">expr</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">samples</span> <span class="ow">is</span> <span class="k">None</span><span class="p">:</span>
        <span class="n">expr_long</span> <span class="o">=</span> <span class="n">expr_long</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">expr_long</span><span class="o">.</span><span class="n">SM</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">samples</span><span class="p">),</span> <span class="p">:]</span>
    <span class="k">if</span> <span class="s">&quot;spikein&quot;</span> <span class="ow">in</span> <span class="n">expr_long</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">expr_long</span> <span class="o">=</span> <span class="n">expr_long</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">expr_long</span><span class="p">[</span><span class="s">&quot;spikein&quot;</span><span class="p">]</span><span class="o">==</span><span class="mi">0</span><span class="p">,]</span>
    <span class="n">expr_long</span><span class="p">[</span><span class="s">&quot;TPM&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">math</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">x</span><span class="o">+</span><span class="mf">1.0</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">expr_long</span><span class="p">[</span><span class="s">&quot;TPM&quot;</span><span class="p">]]</span>
    <span class="n">expr_long</span><span class="p">[</span><span class="s">&quot;FPKM&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">math</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">x</span><span class="o">+</span><span class="mf">1.0</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">expr_long</span><span class="p">[</span><span class="s">&quot;FPKM&quot;</span><span class="p">]]</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">expr_long</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s">&quot;gene_id&quot;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="n">pca_quant</span><span class="p">,</span>
                              <span class="n">index</span><span class="o">=</span><span class="s">&quot;SM&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">screen_n</span> <span class="ow">is</span> <span class="k">None</span><span class="p">:</span>
        <span class="n">smllogger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Only using top {} variable features&quot;</span><span class="p">,</span> <span class="n">screen_n</span><span class="p">)</span>
        <span class="n">sorted_vars</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">var</span><span class="p">())),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="k">True</span><span class="p">)]</span>
        <span class="n">X</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">sorted_vars</span><span class="p">[:</span><span class="n">screen_n</span><span class="p">]]</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">expr_long</span><span class="p">[</span><span class="n">expr_long</span><span class="p">[</span><span class="s">&quot;gene_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)]</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s">&quot;gene_id&quot;</span><span class="p">,</span> <span class="s">&quot;gene_name&quot;</span><span class="p">]]</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;labels&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span><span class="n">v</span> <span class="k">for</span> <span class="p">(</span><span class="n">k</span><span class="p">,</span><span class="n">v</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s">&quot;gene_id&quot;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s">&quot;gene_name&quot;</span><span class="p">])}</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="n">smllogger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;Failed to set labels on gene_name&quot;</span><span class="p">)</span>
    <span class="n">pcares</span><span class="p">,</span> <span class="n">pcaobj</span> <span class="o">=</span> <span class="n">scrnaseq_pca</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="n">detected_genes</span> <span class="o">=</span> <span class="n">expr_long</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;groupby&quot;</span><span class="p">,</span> <span class="s">&#39;SM&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">x</span> <span class="o">&gt;</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&#39;cutoff&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">detected_genes</span> <span class="ow">is</span> <span class="k">None</span><span class="p">:</span>
        <span class="n">pcares</span> <span class="o">=</span> <span class="n">pcares</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">detected_genes</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">pca_output_file</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
        <span class="n">pcares</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">pcaobj_pickle_file</span><span class="p">,</span> <span class="s">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
        <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">pcaobj</span><span class="p">,</span> <span class="n">fh</span><span class="p">)</span></div>


<div class="viewcode-block" id="scrnaseq_pca"><a class="viewcode-back" href="../../../../docs/reference/scrnaseq.html#snakemakelib_workflows.scrnaseq.app.pca.scrnaseq_pca">[docs]</a><span class="k">def</span> <span class="nf">scrnaseq_pca</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s">&quot;PCA&quot;</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="k">None</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="k">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;WIP: Run pca on expression values.</span>

<span class="sd">    Args:</span>
<span class="sd">      X (DataFrame): nxp gene expression matrix, where n=samples, p=features (genes)</span>
<span class="sd">      method (str): what pca method to use</span>
<span class="sd">      labels (list): additional feature labels (e.g. matrix X has gene ids, whereas labels could be gene names)</span>
<span class="sd">      metadata (str): metadata file name for annotation</span>


<span class="sd">    FIXME: generic enough to put in snakemakelib-core?</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s">&quot;sparsePCA&quot;</span><span class="p">:</span>
        <span class="c">### Tuning parameter should be chosen from cross validation</span>
        <span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="k">import</span> <span class="n">sparsePCA</span> <span class="k">as</span> <span class="n">PCA</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="k">import</span> <span class="n">PCA</span>

    <span class="c"># Set multiindex on columns</span>
    <span class="n">pcaobj</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">whiten</span><span class="o">=</span><span class="k">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="n">pcaobj</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
    <span class="n">pcaobj</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">labels</span> <span class="ow">is</span> <span class="k">None</span><span class="p">:</span>
        <span class="n">pcaobj</span><span class="o">.</span><span class="n">labels</span> <span class="o">=</span> <span class="n">labels</span>
    <span class="n">pcares</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pcaobj</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">)</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">X</span><span class="p">))</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">X</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="ow">is</span> <span class="k">None</span><span class="p">:</span>
        <span class="n">pcares</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">index</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">metadata</span> <span class="ow">is</span> <span class="k">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">md</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="n">X</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&quot;expression index name &#39;{name}&#39; not present in &#39;{metadata}&#39;; make sure &#39;{name}&#39; column exists in order to merge pca results with metadata&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">X</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">))</span>
        <span class="n">pcares</span> <span class="o">=</span> <span class="n">pcares</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">md</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pcares</span><span class="p">,</span> <span class="n">pcaobj</span></div>


</pre></div>

    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
        <br/>
        
      
    </p>
    <p>
        &copy; Copyright 2015, Per Unneberg.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.8.<br/>
    </p>
  </div>
</footer>
  </body>
</html>