<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>snakemakelib_workflows.scrnaseq.app.report &#8212; snakemakelib_workflows 0.1-alpha.9+3.gad8e79e documentation</title>
    
    <link rel="stylesheet" href="../../../../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/bootstrap-3.3.4/css/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/bootstrap-3.3.4/css/bootstrap-theme.min.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../../',
        VERSION:     '0.1-alpha.9+3.gad8e79e',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../../../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../../../../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../../../../_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../../../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
    <link rel="top" title="snakemakelib_workflows 0.1-alpha.9+3.gad8e79e documentation" href="../../../../index.html" />
    <link rel="up" title="Module code" href="../../../index.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body role="document">


<div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">

    <div class="container">

      <div class="navbar-header">

        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>

        <a class="navbar-brand" href="../../../../index.html">snakemakelib_workflows</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1-alpha.9+3.gad8e79e</b></span>
      </div>
        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../../../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/workflows.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/workflows.html#general-comments">General comments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/workflows.html#workflow-list">Workflow list</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../docs/workflows/atacseq.html">ATAC-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/atacseq.html#configuration">Configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/atacseq.html#workflow">Workflow</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/atacseq.html#troubleshooting">Troubleshooting</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/atacseq.html#reference-guide">Reference guide</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html">Single cell RNA-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#required-configuration">Required configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#workflow-specific-configuration">Workflow specific configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#other-configuration">Other configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#additional-advice">Additional advice</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#workflow">Workflow</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#troubleshooting">Troubleshooting</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/workflows/scrnaseq.html#reference-guide">Reference guide</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/workflows.html#known-issues">Known issues</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/workflows.html#troubleshooting">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/reference.html">Reference Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../docs/reference/atacseq.html"><code class="docutils literal"><span class="pre">snakemakelib_workflows.atacseq.app</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/atacseq.html#graphics"><code class="docutils literal"><span class="pre">graphics</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/atacseq.html#hooks"><code class="docutils literal"><span class="pre">hooks</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/atacseq.html#results"><code class="docutils literal"><span class="pre">results</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html"><code class="docutils literal"><span class="pre">snakemakelib_workflows.scrnaseq.app</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#graphics"><code class="docutils literal"><span class="pre">graphics</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#hooks"><code class="docutils literal"><span class="pre">hooks</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#pca"><code class="docutils literal"><span class="pre">pca</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#report"><code class="docutils literal"><span class="pre">report</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#technicalnoise"><code class="docutils literal"><span class="pre">technicalnoise</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/reference/scrnaseq.html#utils"><code class="docutils literal"><span class="pre">utils</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs/release_notes.html">Release notes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../docs/release_notes.html#pre-releases">Pre-releases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.9.html">0.1-alpha.9 (Oct 11 2016)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.8.html">0.1-alpha.8 (Jan 8 2016)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.7.html">0.1-alpha.7 (Jan 7 2016)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.6.html">0.1-alpha.6 (Dec 13 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.5.html">0.1-alpha.5 (Dec 12 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.4.html">0.1-alpha.4 (Dec 12 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.3.html">0.1-alpha.3 (Dec 12 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.2.html">0.1-alpha.2 (Nov 24 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../docs/releases/0.1-alpha.1.html">0.1-alpha.1 (Oct 8 2015)</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"></ul>
</li>
              
            
            
            
            
            
          </ul>

        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <h1>Source code for snakemakelib_workflows.scrnaseq.app.report</h1><div class="highlight"><pre>
<span class="c"># Copyright (C) 2015 by Per Unneberg</span>
<span class="kn">import</span> <span class="nn">jinja2</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">snakemake.report</span> <span class="k">import</span> <span class="n">data_uri</span>
<span class="kn">from</span> <span class="nn">snakemakelib.resources</span> <span class="k">import</span> <span class="n">bootstrap_css_files</span><span class="p">,</span> <span class="n">bootstrap_js_files</span>
<span class="kn">from</span> <span class="nn">snakemakelib.plot.bokeh</span> <span class="k">import</span> <span class="n">static_html</span>
<span class="kn">from</span> <span class="nn">snakemakelib.plot.sklearn</span> <span class="k">import</span> <span class="n">plot_pca</span>
<span class="kn">from</span> <span class="nn">snakemakelib</span> <span class="k">import</span> <span class="n">SNAKEMAKELIB_TEMPLATES_PATH</span>
<span class="kn">from</span> <span class="nn">snakemakelib_workflows</span> <span class="k">import</span> <span class="n">SNAKEMAKELIB_WORKFLOWS_TEMPLATES_PATH</span><span class="p">,</span> <span class="n">all_versions</span>
<span class="kn">from</span> <span class="nn">snakemakelib.log</span> <span class="k">import</span> <span class="n">LoggerManager</span>

<span class="n">smllogger</span> <span class="o">=</span> <span class="n">LoggerManager</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="n">__name__</span><span class="p">)</span>

<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;scrnaseq_save_align_rseqc_metrics&#39;</span><span class="p">,</span> <span class="s">&#39;scrnaseq_qc&#39;</span><span class="p">]</span>

<div class="viewcode-block" id="scrnaseq_save_align_rseqc_metrics"><a class="viewcode-back" href="../../../../docs/reference/scrnaseq.html#snakemakelib_workflows.scrnaseq.app.scrnaseq_save_align_rseqc_metrics">[docs]</a><span class="k">def</span> <span class="nf">scrnaseq_save_align_rseqc_metrics</span><span class="p">(</span><span class="n">config</span><span class="p">,</span> <span class="n">output</span><span class="p">,</span> <span class="n">Align</span><span class="p">,</span> <span class="n">RSeQC_readDistribution</span><span class="p">,</span> <span class="n">RSeQC_geneBodyCoverage</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Helper function: join alignment results with rseqc results&quot;&quot;&quot;</span>
    <span class="n">Align</span><span class="o">.</span><span class="n">read_aggregate_data</span><span class="p">(</span><span class="s">&#39;log&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="s">&quot;SM&quot;</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span>  <span class="n">Align</span><span class="o">.</span><span class="n">aggregate_data</span><span class="p">[</span><span class="s">&#39;log&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">df</span><span class="p">[</span><span class="s">&quot;value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s">&quot;value&quot;</span><span class="p">],</span> <span class="n">errors</span><span class="o">=</span><span class="s">&#39;coerce&#39;</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="s">&quot;value&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="s">&quot;SM&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s">&quot;name&quot;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">RSeQC_readDistribution</span><span class="o">.</span><span class="n">read_aggregate_data</span><span class="p">(</span><span class="s">&#39;txt&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="s">&quot;SM&quot;</span><span class="p">)</span>
        <span class="n">RSeQC_geneBodyCoverage</span><span class="o">.</span><span class="n">read_aggregate_data</span><span class="p">(</span><span class="s">&#39;txt&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="s">&quot;SM&quot;</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">RSeQC_readDistribution</span><span class="o">.</span><span class="n">aggregate_data</span><span class="p">[</span><span class="s">&quot;txt&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s">&quot;Group&quot;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s">&quot;Tag_count&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="s">&quot;SM&quot;</span><span class="p">))</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">RSeQC_geneBodyCoverage</span><span class="o">.</span><span class="n">aggregate_data</span><span class="p">[</span><span class="s">&quot;txt&quot;</span><span class="p">])</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="n">smllogger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;rseqc_read_distribution and/or rseqc_geneBodyCoverage not run; only saving alignment results&quot;</span><span class="p">)</span>
    <span class="c"># Remember: ColumnDataSource column names cannot include spaces or %</span>
    <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&quot; &quot;</span><span class="p">,</span> <span class="s">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&quot;%&quot;</span><span class="p">,</span> <span class="s">&quot;PCT&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output</span><span class="o">.</span><span class="n">csv</span><span class="p">)</span></div>


<span class="c"># Template environment</span>
<span class="n">loader</span> <span class="o">=</span> <span class="n">jinja2</span><span class="o">.</span><span class="n">FileSystemLoader</span><span class="p">([</span><span class="n">SNAKEMAKELIB_TEMPLATES_PATH</span><span class="p">,</span>
                                  <span class="n">SNAKEMAKELIB_WORKFLOWS_TEMPLATES_PATH</span><span class="p">])</span>
<span class="n">Env</span> <span class="o">=</span> <span class="n">jinja2</span><span class="o">.</span><span class="n">Environment</span><span class="p">(</span><span class="n">loader</span><span class="o">=</span><span class="n">loader</span><span class="p">)</span>


<div class="viewcode-block" id="scrnaseq_qc"><a class="viewcode-back" href="../../../../docs/reference/scrnaseq.html#snakemakelib_workflows.scrnaseq.app.scrnaseq_qc">[docs]</a><span class="k">def</span> <span class="nf">scrnaseq_qc</span><span class="p">(</span><span class="n">config</span><span class="p">,</span> <span class="nb">input</span><span class="p">,</span> <span class="n">output</span><span class="p">,</span> <span class="n">results</span><span class="o">=</span><span class="k">None</span><span class="p">,</span> <span class="n">rsem</span><span class="o">=</span><span class="k">None</span><span class="p">,</span> <span class="n">rpkmforgenes</span><span class="o">=</span><span class="k">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Do QC of scrnaseq&quot;&quot;&quot;</span>
    <span class="c"># Collect results - subset by samples if applicable</span>
    <span class="n">results</span><span class="o">.</span><span class="n">read_aggregate_data</span><span class="p">()</span>
    <span class="n">results</span><span class="o">.</span><span class="n">filter_aggregate_data</span><span class="p">(</span><span class="s">&#39;alignrseqc&#39;</span><span class="p">,</span> <span class="s">&#39;SM&#39;</span><span class="p">,</span> <span class="n">config</span><span class="p">[</span><span class="s">&quot;samples&quot;</span><span class="p">])</span>

    <span class="n">taptool_url</span> <span class="o">=</span> <span class="k">None</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">config</span><span class="p">[</span><span class="s">&#39;scrnaseq.workflow&#39;</span><span class="p">][</span><span class="s">&#39;report&#39;</span><span class="p">][</span><span class="s">&#39;annotation_url&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="k">None</span><span class="p">:</span>
        <span class="n">taptool_url</span> <span class="o">=</span> <span class="n">config</span><span class="p">[</span><span class="s">&#39;scrnaseq.workflow&#39;</span><span class="p">][</span><span class="s">&#39;report&#39;</span><span class="p">][</span><span class="s">&#39;annotation_url&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;@gene_id&quot;</span>
        
    <span class="c"># Brennecke args</span>
    <span class="c"># brennecke_args = {&#39;plot_height&#39;:600, &#39;plot_width&#39;: 800, &#39;alpha&#39;:</span>
    <span class="c">#                   0.3, &#39;taptool_url&#39;: config[&#39;scrnaseq.workflow&#39;][&#39;report&#39;][&#39;annotation_url&#39;] + &quot;@gene_id&quot;}</span>
    <span class="c"># Alignment stats</span>
    <span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;align&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s">&#39;fig&#39;</span><span class="p">:</span> <span class="n">results</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="s">&#39;alignrseqc&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;fig&#39;</span><span class="p">],</span> <span class="s">&#39;table&#39;</span><span class="p">:</span> <span class="n">results</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="s">&#39;alignrseqc&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;table&#39;</span><span class="p">]}}</span>
    <span class="c"># rsem plots</span>
    <span class="k">if</span> <span class="n">rsem</span><span class="o">.</span><span class="n">run</span><span class="p">:</span>
        <span class="n">d</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s">&#39;rsem&#39;</span><span class="p">:</span> <span class="n">plot_pca</span><span class="p">(</span><span class="n">rsem</span><span class="o">.</span><span class="n">targets</span><span class="p">[</span><span class="s">&#39;pca&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                                   <span class="n">config</span><span class="p">[</span><span class="s">&#39;scrnaseq.workflow&#39;</span><span class="p">][</span><span class="s">&#39;metadata&#39;</span><span class="p">],</span>
                                   <span class="n">rsem</span><span class="o">.</span><span class="n">targets</span><span class="p">[</span><span class="s">&#39;pca&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&quot;.pca.csv&quot;</span><span class="p">,</span> <span class="s">&quot;.pcaobj.pickle&quot;</span><span class="p">),</span>
                                   <span class="n">taptool_url</span><span class="o">=</span> <span class="n">taptool_url</span><span class="p">)})</span>
    <span class="c">#     # FIXME: Instead of re use list</span>
    <span class="c">#     # d[&#39;rsem&#39;].update({&#39;brennecke&#39;: scrnaseq_brennecke_plot(infile=input.rsemgenes, spikein_re=re.compile(&quot;^ERCC&quot;),</span>
    <span class="c">#     #                                                        index=[&quot;SM&quot;, &quot;gene_id&quot;, &quot;transcript_id(s)&quot;, &quot;gene_name&quot;],</span>
    <span class="c">#     #                                                        **brennecke_args)})</span>

    <span class="c"># rpkmforgenes plots</span>
    <span class="k">if</span> <span class="n">rpkmforgenes</span><span class="o">.</span><span class="n">run</span><span class="p">:</span>
        <span class="n">d</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s">&#39;rpkmforgenes&#39;</span><span class="p">:</span> <span class="n">plot_pca</span><span class="p">(</span><span class="n">rpkmforgenes</span><span class="o">.</span><span class="n">targets</span><span class="p">[</span><span class="s">&#39;pca&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                                           <span class="n">config</span><span class="p">[</span><span class="s">&#39;scrnaseq.workflow&#39;</span><span class="p">][</span><span class="s">&#39;metadata&#39;</span><span class="p">],</span>
                                           <span class="n">rpkmforgenes</span><span class="o">.</span><span class="n">targets</span><span class="p">[</span><span class="s">&#39;pca&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&quot;.pca.csv&quot;</span><span class="p">,</span> <span class="s">&quot;.pcaobj.pickle&quot;</span><span class="p">),</span>
                                           <span class="n">taptool_url</span><span class="o">=</span> <span class="n">taptool_url</span><span class="p">)})</span>


    <span class="n">d</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s">&#39;version&#39;</span> <span class="p">:</span> <span class="n">all_versions</span><span class="p">(),</span> <span class="s">&#39;config&#39;</span> <span class="p">:</span> <span class="p">{</span><span class="s">&#39;uri&#39;</span> <span class="p">:</span> <span class="n">data_uri</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">globalconf</span><span class="p">),</span> <span class="s">&#39;file&#39;</span> <span class="p">:</span> <span class="nb">input</span><span class="o">.</span><span class="n">globalconf</span><span class="p">}})</span>
    <span class="n">d</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s">&#39;rulegraph&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s">&#39;fig&#39;</span><span class="p">:</span> <span class="nb">input</span><span class="o">.</span><span class="n">rulegraph</span><span class="p">,</span> <span class="s">&#39;uri&#39;</span><span class="p">:</span> <span class="n">data_uri</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">rulegraph</span><span class="p">),</span>
                            <span class="s">&#39;target&#39;</span><span class="p">:</span> <span class="s">&#39;scrnaseq_all&#39;</span><span class="p">}})</span>
    <span class="n">tp</span> <span class="o">=</span> <span class="n">Env</span><span class="o">.</span><span class="n">get_template</span><span class="p">(</span><span class="s">&#39;workflow_scrnaseq_qc.html&#39;</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output</span><span class="o">.</span><span class="n">html</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
        <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">static_html</span><span class="p">(</span><span class="n">tp</span><span class="p">,</span> <span class="n">template_variables</span><span class="o">=</span><span class="n">d</span><span class="p">,</span> <span class="n">css_raw</span><span class="o">=</span><span class="n">bootstrap_css_files</span><span class="p">,</span> <span class="n">js_raw</span><span class="o">=</span><span class="n">bootstrap_js_files</span><span class="p">))</span></div>
</pre></div>

    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
        <br/>
        
      
    </p>
    <p>
        &copy; Copyright 2015, Per Unneberg.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.8.<br/>
    </p>
  </div>
</footer>
  </body>
</html>