<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Single cell RNA-seq &#8212; snakemakelib_workflows 0.1-alpha.9+1.g63cbc92 documentation</title>
    
    <link rel="stylesheet" href="../../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/bootstrap-3.3.4/css/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/bootstrap-3.3.4/css/bootstrap-theme.min.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.1-alpha.9+1.g63cbc92',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../../_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="top" title="snakemakelib_workflows 0.1-alpha.9+1.g63cbc92 documentation" href="../../index.html" />
    <link rel="up" title="About" href="../workflows.html" />
    <link rel="next" title="snakemakelib_workflows.scrnaseq.app" href="../reference/scrnaseq.html" />
    <link rel="prev" title="snakemakelib_workflows.atacseq.app" href="../reference/atacseq.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body role="document">


<div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">

    <div class="container">

      <div class="navbar-header">

        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>

        <a class="navbar-brand" href="../../index.html">snakemakelib_workflows</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1-alpha.9+1.g63cbc92</b></span>
      </div>
        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../workflows.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflows.html#general-comments">General comments</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../workflows.html#workflow-list">Workflow list</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="atacseq.html">ATAC-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="atacseq.html#configuration">Configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="atacseq.html#workflow">Workflow</a></li>
<li class="toctree-l3"><a class="reference internal" href="atacseq.html#troubleshooting">Troubleshooting</a></li>
<li class="toctree-l3"><a class="reference internal" href="atacseq.html#reference-guide">Reference guide</a></li>
</ul>
</li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Single cell RNA-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#required-configuration">Required configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="#workflow-specific-configuration">Workflow specific configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="#other-configuration">Other configuration</a></li>
<li class="toctree-l3"><a class="reference internal" href="#additional-advice">Additional advice</a></li>
<li class="toctree-l3"><a class="reference internal" href="#workflow">Workflow</a></li>
<li class="toctree-l3"><a class="reference internal" href="#troubleshooting">Troubleshooting</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reference-guide">Reference guide</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../workflows.html#known-issues">Known issues</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflows.html#troubleshooting">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference.html">Reference Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference/atacseq.html"><code class="docutils literal"><span class="pre">snakemakelib_workflows.atacseq.app</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../reference/atacseq.html#graphics"><code class="docutils literal"><span class="pre">graphics</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../reference/atacseq.html#hooks"><code class="docutils literal"><span class="pre">hooks</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../reference/atacseq.html#results"><code class="docutils literal"><span class="pre">results</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../reference/scrnaseq.html"><code class="docutils literal"><span class="pre">snakemakelib_workflows.scrnaseq.app</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../reference/scrnaseq.html#graphics"><code class="docutils literal"><span class="pre">graphics</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../reference/scrnaseq.html#hooks"><code class="docutils literal"><span class="pre">hooks</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../reference/scrnaseq.html#pca"><code class="docutils literal"><span class="pre">pca</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../reference/scrnaseq.html#report"><code class="docutils literal"><span class="pre">report</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../reference/scrnaseq.html#technicalnoise"><code class="docutils literal"><span class="pre">technicalnoise</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../reference/scrnaseq.html#utils"><code class="docutils literal"><span class="pre">utils</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../release_notes.html">Release notes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../release_notes.html#pre-releases">Pre-releases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../releases/0.1-alpha.9.html">0.1-alpha.9 (Oct 11 2016)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../releases/0.1-alpha.8.html">0.1-alpha.8 (Jan 8 2016)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../releases/0.1-alpha.7.html">0.1-alpha.7 (Jan 7 2016)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../releases/0.1-alpha.6.html">0.1-alpha.6 (Dec 13 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../releases/0.1-alpha.5.html">0.1-alpha.5 (Dec 12 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../releases/0.1-alpha.4.html">0.1-alpha.4 (Dec 12 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../releases/0.1-alpha.3.html">0.1-alpha.3 (Dec 12 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../releases/0.1-alpha.2.html">0.1-alpha.2 (Nov 24 2015)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../releases/0.1-alpha.1.html">0.1-alpha.1 (Oct 8 2015)</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Single cell RNA-seq</a><ul>
<li><a class="reference internal" href="#required-configuration">Required configuration</a></li>
<li><a class="reference internal" href="#workflow-specific-configuration">Workflow specific configuration</a></li>
<li><a class="reference internal" href="#other-configuration">Other configuration</a></li>
<li><a class="reference internal" href="#additional-advice">Additional advice</a></li>
<li><a class="reference internal" href="#workflow">Workflow</a></li>
<li><a class="reference internal" href="#troubleshooting">Troubleshooting</a></li>
<li><a class="reference internal" href="#reference-guide">Reference guide</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="single-cell-rna-seq">
<h1>Single cell RNA-seq<a class="headerlink" href="#single-cell-rna-seq" title="Permalink to this headline">¶</a></h1>
<div class="section" id="required-configuration">
<h2>Required configuration<a class="headerlink" href="#required-configuration" title="Permalink to this headline">¶</a></h2>
<p>The following options must be set in the configuration file:</p>
<div class="highlight-yaml"><div class="highlight"><pre><span class="l-Scalar-Plain">bio.ngs.settings</span><span class="p-Indicator">:</span>
  <span class="l-Scalar-Plain">db</span><span class="p-Indicator">:</span>
    <span class="l-Scalar-Plain">ref</span><span class="p-Indicator">:</span> <span class="c1"># Reference sequence</span>
  <span class="l-Scalar-Plain">annotation</span><span class="p-Indicator">:</span>
    <span class="l-Scalar-Plain">transcript_annot_gtf</span><span class="p-Indicator">:</span> <span class="c1"># transcript annotation file in gtf format</span>

<span class="l-Scalar-Plain">samples</span><span class="p-Indicator">:</span> <span class="c1"># list of sample identifiers</span>
</pre></div>
</div>
</div>
<div class="section" id="workflow-specific-configuration">
<h2>Workflow specific configuration<a class="headerlink" href="#workflow-specific-configuration" title="Permalink to this headline">¶</a></h2>
<div class="highlight-yaml"><div class="highlight"><pre><span class="l-Scalar-Plain">scrnaseq.workflow</span><span class="p-Indicator">:</span>
  <span class="l-Scalar-Plain">quantification</span><span class="p-Indicator">:</span> <span class="c1"># One or both of the following</span>
    <span class="p-Indicator">-</span> <span class="l-Scalar-Plain">rsem</span>
    <span class="p-Indicator">-</span> <span class="l-Scalar-Plain">rpkmforgenes</span>
  <span class="l-Scalar-Plain">metadata</span><span class="p-Indicator">:</span> <span class="c1"># Tab-delimited or csv file with additional information about the samples</span>
  <span class="l-Scalar-Plain">report</span><span class="p-Indicator">:</span>
    <span class="l-Scalar-Plain">annotation_url</span><span class="p-Indicator">:</span> <span class="c1"># optional link to gene database, e.g. ensembl</span>
</pre></div>
</div>
</div>
<div class="section" id="other-configuration">
<h2>Other configuration<a class="headerlink" href="#other-configuration" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt><cite>[&#8216;bio.ngs.settings&#8217;][&#8216;db&#8217;][&#8216;extra_ref&#8217;]</cite></dt>
<dd>If you have additional reference sequence to include, such as spikeins
or marker protein sequence, include these in section
[&#8216;bio.ngs.settings&#8217;][&#8216;db&#8217;][&#8216;extra_ref&#8217;]. This parameter takes a list
of fasta file names. Since STAR can take multiple input files for
indexing, the files will not be concatenated.</dd>
</dl>
</div>
<div class="section" id="additional-advice">
<h2>Additional advice<a class="headerlink" href="#additional-advice" title="Permalink to this headline">¶</a></h2>
<p>There are a couple of helper rules for generating spikein input files
and the transcript annotation file.</p>
<dl class="docutils">
<dt><cite>bio.ngs.db.utils.dbutils_make_transcript_annot_gtf</cite></dt>
<dd>For QC statistics calculated by RSEQC, the gtf annotation file
should reflect the content of the alignment index. You can
automatically create the file name defined in
<cite>[&#8216;bio.ngs.settings&#8217;][&#8216;annotation&#8217;][&#8216;transcript_annot_gtf&#8217;]</cite> from
the list of files defined in
<cite>[&#8216;bio.ngs.settings&#8217;][&#8216;annotation&#8217;][&#8216;sources&#8217;]</cite> via the rule
<cite>dbutils_make_transcript_annot_gtf</cite>. gtf and genbank input format is
accepted.</dd>
<dt><cite>bio.ngs.db.ercc.ercc_create_ref</cite></dt>
<dd>The <a class="reference external" href="https://www.thermofisher.com/order/catalog/product/4456740">ERCC RNA Spike-In Mix</a> is
commonly used as spike-in. The rule <cite>ercc_create_ref</cite> automates
download of the sequences in fasta and genbank formats.</dd>
</dl>
</div>
<div class="section" id="workflow">
<h2>Workflow<a class="headerlink" href="#workflow" title="Permalink to this headline">¶</a></h2>
<p>The figure below illustrates the workflow included in the test
directory. Here, use has been made of the additional rules mentioned
above to generate ERCC spike-ins and the transcript annotation gtf.</p>
<img src="../../_images/graphviz-b3ac0fd2cae8f1a4f43a6e481ea38fb6b5c12c05.png" alt="digraph snakemake_dag {
 graph[bgcolor=white, margin=0];
 node[shape=box, style=rounded, fontname=sans,                 fontsize=10, penwidth=2];
 edge[penwidth=2, color=grey];
     0[label = &quot;conf_yaml&quot;, color = &quot;0.27 0.6 0.85&quot;, style=&quot;rounded&quot;];
     1[label = &quot;star_align_log&quot;, color = &quot;0.17 0.6 0.85&quot;, style=&quot;rounded&quot;];
     2[label = &quot;rulegraph_png&quot;, color = &quot;0.02 0.6 0.85&quot;, style=&quot;rounded&quot;];
     3[label = &quot;rseqc_read_quality&quot;, color = &quot;0.28 0.6 0.85&quot;, style=&quot;rounded&quot;];
     4[label = &quot;scrnaseq_aggregate_rseqc_genebody_coverage&quot;, color = &quot;0.29 0.6 0.85&quot;, style=&quot;rounded&quot;];
     5[label = &quot;emboss_seqret_fa_to_genbank&quot;, color = &quot;0.05 0.6 0.85&quot;, style=&quot;rounded&quot;];
     6[label = &quot;picard_sort_sam&quot;, color = &quot;0.06 0.6 0.85&quot;, style=&quot;rounded&quot;];
     7[label = &quot;scrnaseq_qc&quot;, color = &quot;0.62 0.6 0.85&quot;, style=&quot;rounded&quot;];
     8[label = &quot;rseqc_qc&quot;, color = &quot;0.09 0.6 0.85&quot;, style=&quot;rounded&quot;];
     9[label = &quot;scrnaseq_aggregate_align&quot;, color = &quot;0.32 0.6 0.85&quot;, style=&quot;rounded&quot;];
     10[label = &quot;gtf_to_bed12&quot;, color = &quot;0.35 0.6 0.85&quot;, style=&quot;rounded&quot;];
     11[label = &quot;ercc_download_metadata&quot;, color = &quot;0.57 0.6 0.85&quot;, style=&quot;rounded&quot;];
     12[label = &quot;samtools_index&quot;, color = &quot;0.39 0.6 0.85&quot;, style=&quot;rounded&quot;];
     13[label = &quot;rseqc_read_duplication&quot;, color = &quot;0.41 0.6 0.85&quot;, style=&quot;rounded&quot;];
     14[label = &quot;scrnaseq_aggregate_rpkmforgenes&quot;, color = &quot;0.00 0.6 0.85&quot;, style=&quot;rounded&quot;];
     15[label = &quot;rseqc_read_GC&quot;, color = &quot;0.43 0.6 0.85&quot;, style=&quot;rounded&quot;];
     16[label = &quot;rpkmforgenes_from_bam&quot;, color = &quot;0.45 0.6 0.85&quot;, style=&quot;rounded&quot;];
     17[label = &quot;save_align_rseqc_data&quot;, color = &quot;0.48 0.6 0.85&quot;, style=&quot;rounded&quot;];
     18[label = &quot;scrnaseq_all&quot;, color = &quot;0.47 0.6 0.85&quot;, style=&quot;rounded&quot;];
     19[label = &quot;scrnaseq_pca&quot;, color = &quot;0.61 0.6 0.85&quot;, style=&quot;rounded&quot;];
     20[label = &quot;rsem_calculate_expression&quot;, color = &quot;0.13 0.6 0.85&quot;, style=&quot;rounded&quot;];
     21[label = &quot;scrnaseq_aggregate_rseqc_read_distribution&quot;, color = &quot;0.50 0.6 0.85&quot;, style=&quot;rounded&quot;];
     22[label = &quot;rsem_prepare_reference&quot;, color = &quot;0.14 0.6 0.85&quot;, style=&quot;rounded&quot;];
     23[label = &quot;rulegraph&quot;, color = &quot;0.15 0.6 0.85&quot;, style=&quot;rounded&quot;];
     24[label = &quot;picard_merge_sam&quot;, color = &quot;0.52 0.6 0.85&quot;, style=&quot;rounded&quot;];
     25[label = &quot;star_index&quot;, color = &quot;0.53 0.6 0.85&quot;, style=&quot;rounded&quot;];
     26[label = &quot;dbutils_make_transcript_annot_gtf&quot;, color = &quot;0.51 0.6 0.85&quot;, style=&quot;rounded&quot;];
     27[label = &quot;ercc_create_ref&quot;, color = &quot;0.42 0.6 0.85&quot;, style=&quot;rounded&quot;];
     28[label = &quot;rseqc_geneBody_coverage&quot;, color = &quot;0.18 0.6 0.85&quot;, style=&quot;rounded&quot;];
     29[label = &quot;bamtools_filter_unique&quot;, color = &quot;0.56 0.6 0.85&quot;, style=&quot;rounded&quot;];
     30[label = &quot;scrnaseq_aggregate_rsem&quot;, color = &quot;0.21 0.6 0.85&quot;, style=&quot;rounded&quot;];
     31[label = &quot;rseqc_read_distribution&quot;, color = &quot;0.22 0.6 0.85&quot;, style=&quot;rounded&quot;];
     32[label = &quot;rseqc_clipping_profile&quot;, color = &quot;0.58 0.6 0.85&quot;, style=&quot;rounded&quot;];
     33[label = &quot;rseqc_junction_annotation&quot;, color = &quot;0.60 0.6 0.85&quot;, style=&quot;rounded&quot;];
     34[label = &quot;star_align_pe&quot;, color = &quot;0.23 0.6 0.85&quot;, style=&quot;rounded&quot;];
     35[label = &quot;scrnaseq_picard_merge_sam_transcript&quot;, color = &quot;0.24 0.6 0.85&quot;, style=&quot;rounded&quot;];
     36[label = &quot;rseqc_read_NVC&quot;, color = &quot;0.66 0.6 0.85&quot;, style=&quot;rounded&quot;];
     34 -&gt; 1
     23 -&gt; 2
     24 -&gt; 3
     28 -&gt; 4
     24 -&gt; 6
     19 -&gt; 7
     17 -&gt; 7
     2 -&gt; 7
     0 -&gt; 7
     3 -&gt; 8
     28 -&gt; 8
     13 -&gt; 8
     31 -&gt; 8
     32 -&gt; 8
     15 -&gt; 8
     33 -&gt; 8
     36 -&gt; 8
     1 -&gt; 9
     26 -&gt; 10
     6 -&gt; 12
     24 -&gt; 13
     16 -&gt; 14
     24 -&gt; 15
     24 -&gt; 16
     26 -&gt; 16
     21 -&gt; 17
     9 -&gt; 17
     4 -&gt; 17
     16 -&gt; 18
     29 -&gt; 18
     8 -&gt; 18
     20 -&gt; 18
     7 -&gt; 18
     30 -&gt; 19
     14 -&gt; 19
     22 -&gt; 20
     35 -&gt; 20
     31 -&gt; 21
     26 -&gt; 22
     29 -&gt; 24
     26 -&gt; 25
     27 -&gt; 25
     5 -&gt; 26
     27 -&gt; 26
     11 -&gt; 27
     6 -&gt; 28
     10 -&gt; 28
     12 -&gt; 28
     34 -&gt; 29
     20 -&gt; 30
     10 -&gt; 31
     24 -&gt; 31
     24 -&gt; 32
     10 -&gt; 33
     24 -&gt; 33
     25 -&gt; 34
     34 -&gt; 35
     24 -&gt; 36
}" />
</div>
<div class="section" id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="reference-guide">
<h2>Reference guide<a class="headerlink" href="#reference-guide" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="../reference/scrnaseq.html"><code class="docutils literal"><span class="pre">snakemakelib_workflows.scrnaseq.app</span></code></a><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference/scrnaseq.html#graphics"><code class="docutils literal"><span class="pre">graphics</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference/scrnaseq.html#hooks"><code class="docutils literal"><span class="pre">hooks</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference/scrnaseq.html#pca"><code class="docutils literal"><span class="pre">pca</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference/scrnaseq.html#report"><code class="docutils literal"><span class="pre">report</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference/scrnaseq.html#technicalnoise"><code class="docutils literal"><span class="pre">technicalnoise</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference/scrnaseq.html#utils"><code class="docutils literal"><span class="pre">utils</span></code></a></li>
</ul>
</li>
</ul>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
        <br/>
        
<div id="sourcelink">
  <a href="../../_sources/docs/workflows/scrnaseq.txt"
     rel="nofollow">Source</a>
</div>
      
    </p>
    <p>
        &copy; Copyright 2015, Per Unneberg.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.8.<br/>
    </p>
  </div>
</footer>
  </body>
</html>